`annotateChipSeqPeaks`<-
function(chip.seq,transcriptDB=NULL,distanceRange=c(-1e+06,1e+06))
# function(chip.seq,transcriptDB=NULL,species="Hs",distanceRange=c(-1e+06,1e+06),file=NULL)
{
	require(GenomicFeatures)
	colnames(chip.seq)<-c("Chromosome","Start","End","Des","Score")
# 	require(paste("org",species,"eg.db",sep="."),character.only=T)
	
	if(is.null(transcriptDB)) stop("Please provide refGenome platform\n")
	else 
	{
	    if(is.character(transcriptDB))
	      genomeDB<-loadFeatures(transcriptDB)
	    else genomeDB<-transcriptDB
	}
	allchr<-unique(chip.seq[,1])

# 	refseqs = names(as.list(get(paste("org",species,"egREFSEQ2EG",sep="."))))
# 	refseqs = elementMetadata(transcripts(genomeDB))$tx_name
	refseqs = select(RefTable, keys(RefTable, "TXNAME"),cols = c("GENEID","TXNAME"),keytype="TXNAME")
	exps = matrix(NA,ncol=5)
 	colnames(exps)<-c("RefID","GeneID","Chromosome","TSSDist","Score")
# 	colnames(exps)<-c("RefID","Chromosome","TSSDist","Score")

	for(chr in allchr)
	{
	  cat("Annotating ",chr,"\n")
	  genome<-transcripts(genomeDB,val<-list(tx_name=refseqs[,"TXNAME"],tx_chrom=chr))
	  genome.starts=start(genome)
	  genome.ends=end(genome)
# 	  strandRle<-strand(genome)
# 	  genome.strands=as.vector(strandRle)
	  genome.strands=as.vector(strand(genome))
	  genome.0.starts=genome.starts
	  genome.0.ends=genome.ends
	  genome.0.starts[genome.strands=="+"]<-genome.starts[genome.strands=="+"]+distanceRange[1]
	  genome.0.starts[genome.0.starts<0]<-0
	  genome.0.ends[genome.strands=="+"]<-genome.starts[genome.strands=="+"]+distanceRange[2]
	  genome.0.starts[genome.strands=="-"]<-genome.ends[genome.strands=="-"]+distanceRange[1]
	  genome.0.starts[genome.0.starts<0]<-0
	  genome.0.ends[genome.strands=="-"]<-genome.ends[genome.strands=="-"]+distanceRange[2]
 # 	  gene.id = unlist(mget(as.character(values(genome)[,"tx_name"]),get(paste("org",species,"egREFSEQ2EG",sep="."))))
	  gene.id = refseqs$GENEID[match(as.character(values(genome)[,"tx_name"]),refseqs$TXNAME)]


	  tree <- IntervalTree(IRanges(genome.0.starts,genome.0.ends))
	  temp<-chip.seq[chip.seq$Chromosome==chr,]
# 	  tables<-matchMatrix(findOverlaps(query=IRanges(temp$Start,temp$End),tree,type="within",select="all"))
	  tables<-as.matrix(findOverlaps(query=IRanges(temp$Start,temp$End),tree,type="within",select="all"))
#print("Overlapps")
#print(dim(tables))
#if(dim(tables)[1]==1) print(tables)
	  distances<-genome.starts[tables[,2]]
	  distances[is.element(tables[,2],which(genome.strands=="-"))]<-genome.ends[tables[is.element(tables[,2],which(genome.strands=="-")),2]]
	  starts1<-temp[tables[,1],"Start"]
	  starts2<-temp[tables[,1],"End"]
	  distances<-as.integer(abs((starts1+starts2)/2-distances))
	  tables<-cbind(tables,Dist=distances,Score=temp[tables[,1],"Score"])
#if(dim(tables)[1]==1) print(tables)
	  rowID<-paste(values(genome)[tables[,2],"tx_name"],values(genome)[tables[,2],"tx_id"],tables[,3],tables[,1],sep="-")
#if(dim(tables)[1]==1) print(rowID)
	  if(length(rowID)>length(unique(rowID))) cat(chr,"\n")
 	  tables<-data.frame(RefID=values(genome)[tables[,2],"tx_name"],GeneID=gene.id[tables[,2]],Chromosome=as.character(rep(chr,dim(tables)[1])),
	    TSSDist=tables[,3],Score=tables[,4],row.names=rowID,stringsAsFactors=FALSE)
#if(dim(tables)[1]==1) print(tables)
# 	  tables<-data.frame(values(genome)[tables[,2],"tx_name"],as.character(rep(chr,dim(tables)[1])),tables[,c(3,4)],stringsAsFactors=FALSE)
# 	  rownames(tables)<-rowID
#  	  colnames(tables)<-c("RefID","GeneID","Chromosome","TSSDist","Score")
# 	  colnames(tables)<-c("RefID","Chromosome","TSSDist","Score")
# 	  tables<-data.frame(tables,stringsAsFactors=FALSE)
	  exps<-rbind(exps,tables)
	  rm(tables)
	}
	exps<-exps[!is.na(exps[,1]),]
# 	if(!is.null(file)) save(exps,file=file)
	exps
}
